extensions/markdown/src/extension.ts

```diff
@@ -265,4 +266,33 @@ class MDDocumentContentProvider implements TextDocumentContentProvider {
 			}, 300);
 		}
 	}
+}
+
+class DocumentHeadingsProvider implements vscode.DocumentSymbolProvider {
+
+	// http://daringfireball.net/projects/markdown/syntax#header
+	private static _atxPattern = /^(#){1,6}\s+.+(\s+\1)?/;
```

Comment:
Also, there are a few other cases to consider, such header like text inside of code blocks. These should not be picked up. Probably not essential for a first pass, but worth considering.